ra/ra.go

```diff
@@ -936,259 +952,232 @@ func (ra *RegistrationAuthorityImpl) failOrder(
+		logEvent.NotAfter = cert.NotAfter
+		beeline.AddFieldToTrace(ctx, "cert.not_after", cert.NotAfter)
+
 		result = "successful"
 	}
+
 	logEvent.ResponseTime = ra.clk.Now()
 	ra.log.AuditObject(fmt.Sprintf("Certificate request - %s", result), logEvent)
-	return cert, err
+
+	// Return both the order and the error: if issueCertificateInner worked, then
+	// err will be nil; if it didn't, then we'll propagate that error upwards.
+	return order, err
 }
 
-// issueCertificateInner handles the heavy lifting aspects of certificate
-// issuance.
-//
-// This function is responsible for ensuring that we never try to issue a final
-// certificate twice for the same precertificate, because that has the potential
-// to create certificates with duplicate serials. For instance, this could
-// happen if final certificates were created with different sets of SCTs. This
-// function accomplishes that by bailing on issuance if there is any error in
-// IssueCertificateForPrecertificate; there are no retries, and serials are
-// generated in IssuePrecertificate, so serials with errors are dropped and
-// never have final certificates issued for them (because there is a possibility
-// that the certificate was actually issued but there was an error returning
-// it).
-func (ra *RegistrationAuthorityImpl) issueCertificateInner(
+// validateFinalizeRequest checks that a FinalizeOrder request is fully correct
+// and ready for issuance.
+func (ra *RegistrationAuthorityImpl) validateFinalizeRequest(
 	ctx context.Context,
-	csr *x509.CertificateRequest,
-	acctID accountID,
-	oID orderID,
-	issuerNameID issuance.IssuerNameID,
-	logEvent *certificateRequestEvent) (core.Certificate, error) {
-	emptyCert := core.Certificate{}
-	if acctID <= 0 {
-		return emptyCert, berrors.MalformedError("invalid account ID: %d", acctID)
+	req *rapb.FinalizeOrderRequest,
+	logEvent *certificateRequestEvent) (*x509.CertificateRequest, error) {
+	if req.Order.Id <= 0 {
+		return nil, berrors.MalformedError("invalid order ID: %d", req.Order.Id)
+	}
+
+	if req.Order.RegistrationID <= 0 {
+		return nil, berrors.MalformedError("invalid account ID: %d", req.Order.RegistrationID)
+	}
+
+	if core.AcmeStatus(req.Order.Status) != core.StatusReady {
+		return nil, berrors.OrderNotReadyError(
+			"Order's status (%q) is not acceptable for finalization",
+			req.Order.Status)
 	}
 
-	if oID <= 0 {
-		return emptyCert, berrors.MalformedError("invalid order ID: %d", oID)
+	// There should never be an order with 0 names at the stage, but we check to
+	// be on the safe side, throwing an internal server error if this assumption
+	// is ever violated.
+	if len(req.Order.Names) == 0 {
+		return nil, berrors.InternalServerError("Order has no associated names")
 	}
 
-	regPB, err := ra.SA.GetRegistration(ctx, &sapb.RegistrationID{Id: int64(acctID)})
+	// Parse the CSR from the request
+	csr, err := x509.ParseCertificateRequest(req.Csr)
 	if err != nil {
-		return emptyCert, err
+		return nil, err
```

Comment:
I don't understand the purpose of this conditional.

> If error is not nil, then return nil and an error. 

So in an error scenario, the call stack would go
`validateFinalizeRequest` -> `FinalizeOrder` -> someplace over in `ra_grpc.pb.go` -> the beyond